client/src/test/java/io/getstream/chat/android/client/parser/adapters/CustomObjectAdapterTest.kt

```diff
@@ -0,0 +1,74 @@
+package io.getstream.chat.android.client.parser.adapters
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonParser
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.parser.ChatParserImpl
+import org.amshove.kluent.shouldBeEqualTo
+import org.intellij.lang.annotations.Language
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+
+class CustomObjectAdapterTest {
+    private val parser = ChatParserImpl()
+
+    @BeforeEach
+    fun `check prerequisites`() {
+        val jsonObject = JsonParser.parseString(json).asJsonObject
+        checkJsonStructureAndValues(jsonObject)
+    }
+
+    @Test
+    fun `should deserialize json object with custom fields`() {
+        val message = parser.fromJson(json, Message::class.java)
+
+        message.extraData.apply {
+            get("extraData") shouldBeEqualTo mapOf("key1" to "value1", "key2" to true, "key3" to mapOf("key4" to "val4"))
+            get("customKey1") shouldBeEqualTo "customVal1"
+            get("customKey2") shouldBeEqualTo "customVal2"
+            get("customKey3") shouldBeEqualTo true
+            get("customKey4") shouldBeEqualTo listOf("a", "b", "c")
+        }
+    }
+
+    @Test
+    fun `should serialize json object with custom fields`() {
+        val jsonString: String = parser.toJson(testMessage)
+        val jsonObject = JsonParser.parseString(jsonString).asJsonObject
+        checkJsonStructureAndValues(jsonObject)
+    }
+
+    private fun checkJsonStructureAndValues(jsonObject: JsonObject) {
+        jsonObject.apply {
+            get("id").asString shouldBeEqualTo "8584452-6d711169-0224-41c2-b9aa-1adbe624521b"
+            get("extraData").asJsonObject.apply {
```

Comment:
All this nesting of `apply` calls usually gets messy, but in this case it's actually super readable. Really nicely done! 👍